fasteners/threads.scad

```diff
@@ -1,245 +1,407 @@
+// depth 1 mm.
+//metric_thread (diameter=20, pitch=8, length=16, square=true, thread_size=6,
+//               groove=true, rectangle=0.333);
+
+// English: 1/4 x 20.
+//english_thread (diameter=1/4, threads_per_inch=20, length=1);
+
+// Tapered.  Example -- pipe size 3/4" -- per:
+// http://www.engineeringtoolbox.com/npt-national-pipe-taper-threads-d_750.html
+// english_thread (diameter=1.05, threads_per_inch=14, length=3/4, taper=1/16);
+
+// Thread for mounting on Rohloff hub.
+//difference () {
+//   cylinder (r=20, h=10, $fn=100);
+//
+//   metric_thread (diameter=34, pitch=1, length=10, internal=true, n_starts=6);
+//}
+
+
+// ----------------------------------------------------------------------------
+function segments (diameter) = min (50, max (ceil (diameter*6), 25));
+
+
+// ----------------------------------------------------------------------------
+// diameter -    outside diameter of threads in mm. Default: 8.
+// pitch    -    thread axial "travel" per turn in mm.  Default: 1.
+// length   -    overall axial length of thread in mm.  Default: 1.
+// internal -    true = clearances for internal thread (e.g., a nut).
+//               false = clearances for external thread (e.g., a bolt).
+//               (Internal threads should be "cut out" from a solid using
+//               difference ()).  Default: false.
+// n_starts -    Number of thread starts (e.g., DNA, a "double helix," has
+//               n_starts=2).  See wikipedia Screw_thread.  Default: 1.
+// thread_size - (non-standard) axial width of a single thread "V" - independent
+//               of pitch.  Default: same as pitch.
+// groove      - (non-standard) true = subtract inverted "V" from cylinder
+//                (rather thanadd protruding "V" to cylinder).  Default: false.
+// square      - true = square threads (per
+//               https://en.wikipedia.org/wiki/Square_thread_form).  Default:
+//               false.
+// rectangle   - (non-standard) "Rectangular" thread - ratio depth/(axial) width
+//               Default: 0 (standard "v" thread).
+// angle       - (non-standard) angle (deg) of thread side from perpendicular to
+//               axis (default = standard = 30 degrees).
+// taper       - diameter change per length (National Pipe Thread/ANSI B1.20.1
+//               is 1" diameter per 16" length). Taper decreases from 'diameter'
+//               as z increases.  Default: 0 (no taper).
+// leadin      - 0 (default): no chamfer; 1: chamfer (45 degree) at max-z end;
+//               2: chamfer at both ends, 3: chamfer at z=0 end.
+// leadfac     - scale of leadin chamfer length (default: 1.0 = 1/2 thread).
+// test        - true = do not render threads (just draw "blank" cylinder).
+//               Default: false (draw threads).
+module metric_thread (diameter=8, pitch=1, length=1, internal=false, n_starts=1,
+                      thread_size=-1, groove=false, square=false, rectangle=0,

```

Comment:
I disagree with putting `square` and `rectangle` as parameters to `metric_thread` -- there are no square or rectangle metric threads, afaik.

Reply:
https://en.wikipedia.org/wiki/Square_thread_form

"Square threads are defined as follows by ISO standards:

    Sq 60 × 9 

where Sq designates a square thread, 60 is the nominal diameter in millimeters, and 9 is the pitch in millimeters. When there is no suffix it is a single start thread. If there is a suffix then the value after the multiplication sign is the lead and the value in the parentheses is the pitch. For example:

    Sq 60 × 18 ( P9 ) LH 

would denoted two starts, as the lead divided by the pitch is two. The "LH" denotes a left hand thread."

& pages 3-4 of [this](https://nptel.ac.in/content/storage2/courses/112105125/pdf/module-6%20lesson-1.pdf).

ISO standard and specified in millimeters, I'd say it is metric.

I'm not sure re rectangular, but I think it is nomenclature where the depth and pitch are not symmetrical.

Reply:
Hmm, I stand corrected. I still don't think they should be part of the same module though.